File: src/effects/sparkle.rs
use std::time::Duration;
use rand::Rng;
use crate::effects::{Effect, Buffer};

pub struct SparkleEffect {
    brightness: Vec<f32>,
    target: Vec<f32>,
    num_pixels: usize,
}

impl SparkleEffect {
    pub fn new(num_pixels: usize) -> Self {
        Self {
            brightness: vec![0.0; num_pixels],
            target: vec![0.0; num_pixels],
            num_pixels,
        }
    }
}

impl Effect for SparkleEffect {
    fn name(&self) -> &'static str { "Sparkle" }

    fn update(&mut self, buffer: &mut Buffer, delta: Duration) -> bool {
        let dt = delta.as_secs_f32();
        let mut rng = rand::thread_rng();

        // randomly trigger new sparkles
        let trigger_count = (self.num_pixels as f32 * dt * 3.0) as usize + 1;
        for _ in 0..trigger_count {
            let idx = rng.gen_range(0..self.num_pixels);
            if self.target[idx] < 0.1 {
                self.target[idx] = rng.gen_range(0.5..1.0);
            }
        }

        for i in 0..self.num_pixels {
            let diff = self.target[i] - self.brightness[i];
            if diff.abs() < 0.01 {
                if self.target[i] > 0.5 {
                    self.target[i] = 0.0;
                }
                self.brightness[i] = self.target[i];
            } else {
                self.brightness[i] += diff * dt * 8.0;
            }

            let b = (self.brightness[i] * 255.0) as u8;
            buffer[i] = [b, b, b];
        }

        true
    }
}

File: effects/sparkle.lua
name = "Sparkle"
description = "Pixels independently ramp to a random peak brightness then fade back to dark"

-- Per-pixel brightness with smooth ramp-up then ramp-down, matching Rust SparkleEffect.
local brightness = {}
local target = {}

function init(n)
    brightness = {}
    target = {}
    for i = 1, n do
        brightness[i] = 0.0
        target[i] = 0.0
    end
end

function update(buf, dt)
    local n = buf:len()

    -- Randomly trigger new sparkles
    local triggers = math.floor(n * dt * 3.0) + 1
    for _ = 1, triggers do
        local idx = math.random(1, n)
        if target[idx] < 0.1 then
            target[idx] = math.random() * 0.5 + 0.5
        end
    end

    for i = 1, n do
        local diff = target[i] - brightness[i]
        if math.abs(diff) < 0.01 then
            if target[i] > 0.5 then
                target[i] = 0.0
            end
            brightness[i] = target[i]
        else
            brightness[i] = brightness[i] + diff * dt * 8.0
        end
        local v = math.floor(brightness[i] * 255)
        buf:set(i - 1, v, v, v)
    end

    return true
end
